EngineLayer/PeptideSpectralMatch.cs

```diff
@@ -268,6 +277,117 @@ public static int GetLongestIonSeriesBidirectional(Dictionary<PeptideWithSetModi
             return maxDiffs.Max();
         }
 
+        /// <summary>
+        /// Determine the Fragment Coverage the PSM
+        /// Assigns fragment coverage indices for the PSM and the protein based on Amino Acid Position in Matched Ion Fragments
+        /// </summary>
+        public void GetAminoAcidCoverage()
+        {
+            if (this.BaseSequence != null && this.MatchedFragmentIons != null && this.MatchedFragmentIons.Any() &&
+                this.PeptideLength > 0)
+            {
+                //Pull C terminal and N terminal Fragments and amino acid numbers
+                var nTermFragmentAAPositions = this.MatchedFragmentIons.Where(p =>
+                        p.NeutralTheoreticalProduct.Terminus == FragmentationTerminus.N)
+                    .Select(j => j.NeutralTheoreticalProduct.AminoAcidPosition).Distinct().ToList();
+
+                var cTermFragmentAAPositions = this.MatchedFragmentIons.Where(p =>
+                        p.NeutralTheoreticalProduct.Terminus == FragmentationTerminus.C)
+                    .Select(j => j.NeutralTheoreticalProduct.AminoAcidPosition).Distinct().ToList();
+
+
+                //Create a hashset to store the covered amino acid positions
+                HashSet<int> fragmentCoveredAminoAcids = new();
+
+
+                //Check N term frags first
+                if (nTermFragmentAAPositions.Count > 0)
+                {
+                    nTermFragmentAAPositions.Sort();
+
+                    //if the final NFragment is present, last AA is covered
+                    if (nTermFragmentAAPositions.Contains(this.BaseSequence.Length - 1))
+                    {
+                        fragmentCoveredAminoAcids.Add(this.BaseSequence.Length);
+                    }
+
+                    // if the first NFragment is present, first AA is covered
+                    if (nTermFragmentAAPositions.Contains(1))
+                    {
+                        fragmentCoveredAminoAcids.Add(1);
+                    }
+
+                    //Check all amino acids except for the last one in the list
+                    for (int i = 0; i < nTermFragmentAAPositions.Count - 1; i++)
+                    {
+                        //sequential AA, second one is covered
+                        if (nTermFragmentAAPositions[i + 1] - nTermFragmentAAPositions[i] == 1)
+                        {
+                            fragmentCoveredAminoAcids.Add(nTermFragmentAAPositions[i + 1]);
+                        }
+
```

Comment:
can these be as "elseif" clauses. I mean, if the position is covered, there is no need to see if its covered another way. just do additonal comparisons if not covered.

Reply:
These have to be several if statements, because we are actually checking two positions for complementary coverage. Either before or after the fragment currently being evaluated. They could both be covered, just one, or neither.